apis/iam/v1alpha1/serviceaccountkey_types.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+)
+
+// ServiceAccountKeyParameters defines parameters for a desired IAM ServiceAccountKey
+// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts.keys
+//
+type ServiceAccountKeyParameters struct {
+	// KeyAlgorithm is an optional user-specified string that specifies the type of key and algorithm
+	// to use for the key. The default is currently a 2048-bit RSA key. However this may change in the future.
+	// Possible values:
+	//   "KEY_ALG_UNSPECIFIED" - Not specified.
+	//   "KEY_ALG_RSA_1024" - 1024-bit RSA key
+	//   "KEY_ALG_RSA_2048" - 2048-bit RSA key
+	// +optional
+	// +immutable
+	KeyAlgorithm *string `json:"keyAlgorithm,omitempty"`
+
+	// PrivateKeyType is an optional specification of the output format of the generated private key.
+	// The default value is TYPE_GOOGLE_CREDENTIALS_FILE, which corresponds to the Google Credentials File Format.
+	// Possible values:
+	//   "TYPE_UNSPECIFIED" - Not specified. Equivalent to TYPE_GOOGLE_CREDENTIALS_FILE.
+	//   "TYPE_PKCS12_FILE" - Private key stored in a RFC7292 PKCS #12 document. Password for the PKCS #12 document is "notasecret".
+	//   "TYPE_GOOGLE_CREDENTIALS_FILE" - Google Credentials File format.
+	// +optional
+	// +immutable
+	PrivateKeyType *string `json:"privateKeyType,omitempty"`
+
+	// PublicKeyType is an optional specification of the output format for the associated public key.
+	// The default value is TYPE_RAW_PUBLIC_KEY.
+	// Possible values:
+	//   "TYPE_NONE" - Not specified. Public key is not retrieved via Google Cloud API.
+	//   "TYPE_X509_PEM_FILE" - X509 PEM format.
+	//   "TYPE_RAW_PUBLIC_KEY" - Raw public key.
+	// +optional
+	// +kubebuilder:default=TYPE_RAW_PUBLIC_KEY
+	PublicKeyType *string `json:"publicKeyType,omitempty"`
```

Comment:
```suggestion
	PublicKeyType string `json:"publicKeyType,omitempty"`
```

I think it's safe to type it as `string` since it will never be empty due to having a default.

Reply:
Hmm interesting, `TYPE_NONE` seems to be advised for that purpose. If the empty value means `TYPE_NONE`, how did we come up with the default of `+kubebuilder:default=TYPE_RAW_PUBLIC_KEY`? Should we not default or maybe use `TYPE_NONE` as default here?

Reply:
I think if we'd like to go with a default, it should be the one the provider chooses because of high fidelity representation of the API. So we could possibly have `TYPE_NONE` as default or not have a default and let late-initialization take care of updating it, which is something I just noticed; we don't have late-initialization in this PR. Could you add it for non-required fields?